### PR TITLE
feat: Add post validation rules

### PR DESCRIPTION
**Pull request by ivasilov**
This PR adds a function to add rules which are run after all individual environment checks are run. This will enable checks on the env config as a whole (for example, to check if some variables are incompatible between each other).

Tests and documentation included.
Also added prettier and vscode configs.

### CODE REVIEW

File: src/index.spec.ts [hunk 1]
```diff
@@ -84,15 +80,30 @@ it('should throw if the validation fails', () => {
...
+  const envaridator = new Envaridator();
+  envaridator.registerPostValidation(
+    'THROW_RULE',
+    () => {
+      throw new Error('This rule should throw');
+    },
+    'This rule will throw when the envaridator validate is run',
+  );
+
+  expect(() => envaridator.validate()).toThrow(
```
Comment (spion-h4) [inline src/index.spec.ts:128]: PostValidation shouldn't say "the following environment variables are invalid"

File: src/index.spec.ts [hunk 2]
```diff
@@ -84,15 +80,30 @@ it('should throw if the validation fails', () => {
   );
 });
 
-it('should describe all registered variables', () => {
+it('should throw if a post validation rule fails', () => {
+  const envaridator = new Envaridator();
+  envaridator.registerPostValidation(
```
Comment (spion-h4) [inline src/index.spec.ts:101]: Its best for post-validation rules not to have a name,

File: README.md [hunk 3]
```diff
@@ -5,43 +5,67 @@
...
+);
+
+const newDBURL = envaridator.register(
+  'DATABASE_URL',
+  toi.optional().and(toix.str.url({ protocol: 'postgres' })),
+  'The SQL database url. Must be a PostgreSQL database.',
+);
+
+envaridator.registerPostValidation(
+  'DATABASE_URL_MIGRATION',
```
Comment (spion-h4) [inline README.md]:
```suggestion
'Either DB_URL or DATABASE_URL needs to defined.',
```

File: src/index.spec.ts [hunk 4]
```diff
@@ -59,31 +37,29 @@ it('should throw validation error if Envar value is invalid', () => {
...
   const envaridator = new Envaridator();
 
   envaridator.register(
     'CORRECT_URL',
     toi.required().and(toix.str.url({ protocol: 'https:' })),
-    'A correct url for this app'
+    'A correct url for this app',
   );
 
+  envaridator.registerPostValidation('EXAMPLE_RULE', () => {}, 'Empty post validation rule');
```
Comment (spion-h4) [inline src/index.spec.ts]:
```suggestion
  envaridator.registerPostValidation(() => {}, 'Empty post validation rule');
```

File: README.md [hunk 5]
```diff
@@ -5,43 +5,67 @@
...
+  () => {
+    // if neither DB_URL or DATABASE_URL is defined, throw an error
+    if (!(oldDBURL.value || newDBURL.value)) {
+      throw new Error();
+    }
+  },
+  'Either DB_URL or DATABASE_URL needs to defined.',
```
Comment (spion-h4) [inline README.md]:
```suggestion
```

File: src/index.spec.ts [hunk 6]
```diff
@@ -96,58 +72,84 @@ it('should throw if the validation fails', () => {
...
+    },
+    'This rule will throw when the envaridator validate is run',
   );
-  envaridator.register(
-    'VAR_2',
-    toi.required(),
-    'The second variable. Must be configured correctly.'
+
+  expect(() => envaridator.validate()).toThrow(
+    'The following environment variables are invalid:\n\nTHROW_RULE - This rule should throw',
```
Comment (spion-h4) [inline src/index.spec.ts]:
```suggestion
    'The following post-validation rules failed:\n\nThis rule should throw',
```

File: src/index.spec.ts [hunk 7]
```diff
@@ -96,58 +72,84 @@ it('should throw if the validation fails', () => {
...
 
-it('should describe all registered variables', () => {
+it('should throw if a post validation rule fails', () => {
   const envaridator = new Envaridator();
-  envaridator.register(
-    'VAR_1',
-    toi.required(),
-    'The first variable. Must be configured correctly.'
+  envaridator.registerPostValidation(
+    'THROW_RULE',
```
Comment (spion-h4) [inline src/index.spec.ts]:
```suggestion
```

File: src/index.spec.ts [hunk 8]
```diff
@@ -96,58 +72,84 @@ it('should throw if the validation fails', () => {
...
   );
-  envaridator.register(
-    'VAR_3',
-    toi.required(),
-    'The third variable. Must be configured correctly.'
+});
+
+it('should throw if a post validation rule already exists', () => {
+  const envaridator = new Envaridator();
+  envaridator.registerPostValidation('DEFINED_RULE', () => {}, '');
```
Comment (spion-h4) [inline src/index.spec.ts]:
```suggestion
  envaridator.registerPostValidation(() => {}, '');
```

File: src/index.spec.ts [hunk 9]
```diff
@@ -96,58 +72,84 @@ it('should throw if the validation fails', () => {
...
+
+  expect(() => envaridator.registerPostValidation('DEFINED_RULE', () => {}, '')).toThrow(
```
Comment (spion-h4) [inline src/index.spec.ts]:
```suggestion
  expect(() => envaridator.registerPostValidation(() => {}, '')).toThrow(
```

File: src/index.spec.ts [hunk 10]
```diff
@@ -96,58 +72,84 @@ it('should throw if the validation fails', () => {
...
+  expect(() => envaridator.validate()).toThrow(
+    'The following environment variables are invalid:\n\nTHROW_RULE - This rule should throw',
   );
-  envaridator.register(
-    'VAR_3',
-    toi.required(),
-    'The third variable. Must be configured correctly.'
+});
+
+it('should throw if a post validation rule already exists', () => {
```
Comment (spion-h4) [inline src/index.spec.ts]: they can be added in a list rather than by key

File: src/index.spec.ts [hunk 11]
```diff
@@ -96,58 +72,84 @@ it('should throw if the validation fails', () => {
...
+"#Variables
+**VAR_1** - The first variable. Must be configured correctly.
+**VAR_2** - The second variable. Must be configured correctly.
+**VAR_3** - The third variable. Must be configured correctly."
+`);
+});
+
+it('should describe all rules with some markdown formatting', () => {
+  const envaridator = new Envaridator();
+  envaridator.registerPostValidation('RULE_1', () => {}, 'The first rule. Must be configured correctly.');
```
Comment (spion-h4) [inline src/index.spec.ts]:
```suggestion
  envaridator.registerPostValidation(() => {}, 'The first rule. Must be configured correctly.');
```

File: src/index.spec.ts [hunk 12]
```diff
@@ -96,58 +72,84 @@ it('should throw if the validation fails', () => {
...
+**RULE_1** - The first rule. Must be configured correctly."
+`);
+});
+
+it('should describe all registered variables and rules with some markdown formatting', () => {
+  const envaridator = new Envaridator();
+  envaridator.register('VAR_1', toi.required(), 'The first variable. Must be configured correctly.');
+  envaridator.register('VAR_2', toi.required(), 'The second variable. Must be configured correctly.');
+  envaridator.register('VAR_3', toi.required(), 'The third variable. Must be configured correctly.');
+  envaridator.registerPostValidation('RULE_1', () => {}, 'The first rule. Must be configured correctly.');
```
Comment (spion-h4) [inline src/index.spec.ts]:
```suggestion
  envaridator.registerPostValidation(() => {}, 'The first rule. Must be configured correctly.');
```

File: src/index.ts [hunk 13]
```diff
@@ -32,15 +34,29 @@ export class Envaridator {
...
 
+  /**
+   * registerPostValidation adds a validation rule which is run after all variable validations.
+   * It should be used to validate rules on the whole config i.e. for migrating of an environment variable.
+   *
+   * @param name the name of the environment variable
+   * @param validator function which should throw if the config is not validated
+   * @param description string describing the variable
+   */
+  registerPostValidation(name: string, validator: PostValidator, description: string) {
```
Comment (spion-h4) [inline src/index.ts]:
```suggestion
  registerPostValidation(validator: PostValidator, description: string) {
```

File: src/index.spec.ts (hunk 2 above)
Comment (ivasilov) [reply src/index.spec.ts:101]: Fixed. Removed the name, put the description on first parameter position.

File: src/index.spec.ts (hunk 1 above)
Comment (ivasilov) [reply src/index.spec.ts:128]: Fixed.

File: src/index.ts [hunk 14]
```diff
@@ -61,27 +79,65 @@ export class Envaridator {
...
+        } catch (validationError) {
+          failedRules.push(validationError.message);
+        }
+      });
+    }
+
+    let result = '';
     if (failedVariables.length > 0) {
-      throw new Error(this.status(failedVariables));
+      result += this.variablesStatus(failedVariables);
```
Comment (ghost) [inline src/index.ts:96]: Not sure if you need += here.

**Comment by ghost** [comment]
The program runs as expected. I tried analyzing the code and I can say that I don't have any major concerns about the code itself. I'm approving this.

One note: you can also remove the `examples` folder since it has only one file that contains an example usage of the program, which is already covered in the `README`. We can also do that in another PR.

File: src/index.spec.ts (hunk 1 above)
Comment (spion-h4) [reply src/index.spec.ts:128]: but the test isn't updated?

Comment (spion-h4) [reply src/index.spec.ts:128]: oh i see it, never mind.

**Review by spion-h4 — APPROVED** [review]

**Pull request merged**